cmd/zc_traverser_local.go

```diff
@@ -68,6 +69,46 @@ func (t *localTraverser) getInfoIfSingleFile() (os.FileInfo, bool, error) {
 	return fileInfo, true, nil
 }
 
+func UnfurlSymlinks(symlinkPath string) (result string, err error) {
+	unfurlingPlan := []string{symlinkPath}
+
+	for len(unfurlingPlan) > 0 {
+		item := unfurlingPlan[0]
+
+		fi, err := os.Lstat(item)
+
+		if err != nil {
+			return item, err
+		}
+
+		if fi.Mode()&os.ModeSymlink != 0 {
+			result, err := os.Readlink(item)
+
+			if err != nil {
+				return result, err
+			}
+
+			// We need to discern whether the item symlinked is under the same directory as the symlink.
+			// This is because when this happens, we only get a relative path, not a full path.
+			// So we'll check for the existence of the item under the relative path, and otherwise treat it as a full path.
+			sameDir := filepath.Join(filepath.Dir(item), result)
+			if _, err = os.Stat(sameDir); err == nil {
```

Comment:
Two thoughts/questions about this line:

1. I'm a bit surprised that the call is to Stat. I would have (just intuitively) expected LStat here.  In case it's a symlink to a symlink.  In that case, I'm not totally sure the current code would be wrong. But it does feel odd. What do you think?

2. Is there anything at the start of the path, that is returned by ReadLink, that would tell us if this is, syntactically, a relative path. E.g. does it start with "."?  Asking because I feel a bit uncomfortable with just appending and assuming its relative if the append points to something that exists. It's theoretically possible that appending an absolute (non-relatev) path may also point to something that exists, almost by chance. E.g. we are in //foo/bar and ReadLink returns //aaa/bbb.  If a file or directory //foo/bar/aaa/bbb happens to exist, I'm pretty sure that we will then mistakenly thing that the path was relative. (I think the double //'s before the aaa will be treated as a single / by Windows, but I could be wrong on that).

What about calling filepath.IsAbs() on the value returned by ReadLink, instead of probing for existence?  Would that work?

Reply:
I don't entirely agree with the safeness of this either.

I'm going to do a bit of testing and see if I can syntactically identify a relative path without too much pain. I recall that it felt tricky and like a bit of a complication, however.

Reply:
I can't verify this was ever super necessary, oddly. I must've had to do it for some reason.